infra/token-transfer-server/src/lib/lockup.js

```diff
@@ -19,14 +25,50 @@ const { encryptionSecret, clientUrl } = require('../config')
  * Adds a lockup
  * @param userId - user id of the user adding the lockup
  * @param amount - the amount to be locked
+ * @param early - whether this is an early lockup of the next vest
+ * @param data - additional data to be stored with the lockup, e.g. fingerprint
  * @returns {Promise<Lockup>} Lockup object.
  */
-async function addLockup(userId, amount, data = {}) {
-  const user = await hasBalance(userId, amount)
+async function addLockup(userId, amount, early, data = {}) {
+  // Check if user has sufficient balance
```

Comment:
There is a lock in the controller so this can't get called concurrently, did you mean in addition to that? Not sure if you saw that

Ref https://github.com/OriginProtocol/origin/blob/55c8f5803e3d9708a7a826a558cab712c3b5b008/infra/token-transfer-server/src/controllers/lockup.js#L84

Reply:
Ah sorry I had missed that... yes, AyncLock achieves the same thing as locking for update the user row in the DB, as long as we are certain there will never be more than 1 instance of the T3 server.